File: src/Web/JavaScriptJS/patterns/mediator.js
const assert = require("node:assert/strict");

/** Coordinates colleagues through a mediator instead of direct coupling. */
function run() {
  const events = [];
  const colleagues = new Map([
    ["inventory", (sender, message) => events.push(`inventory<-${sender}:${message}`)],
    ["payment", (sender, message) => events.push(`payment<-${sender}:${message}`)],
  ]);

  const mediator = {
    send(sender, recipient, message) {
      const receiver = colleagues.get(recipient);
      if (!receiver) throw new Error(`Unknown colleague: ${recipient}`);
      receiver(sender, message);
    },
  };

  const payment = {
    send(message) {
      mediator.send("payment", "inventory", message);
    },
  };

  const inventory = {
    send(message) {
      mediator.send("inventory", "payment", message);
    },
  };

  payment.send("paid");
  inventory.send("reserved");

  assert.deepEqual(events, [
    "inventory<-payment:paid",
    "payment<-inventory:reserved",
  ]);
  assert.throws(
    () => mediator.send("payment", "shipping", "paid"),
    /Unknown colleague: shipping/,
  );

  return "mediator";
}

module.exports = { run };
if (require.main === module) run();

File: src/Scripting/Ruby/patterns/mediator.rb
# frozen_string_literal: true

module RubyPatterns
  module Mediator
    class CheckoutMediator
      def initialize
        @colleagues = {}
      end

      def register(name, &receiver)
        @colleagues.fetch(name) { @colleagues[name] = receiver }
      end

      def send(sender, recipient, message)
        @colleagues.fetch(recipient).call(sender, message)
      end
    end

    module_function

    def run
      events = []
      mediator = CheckoutMediator.new
      mediator.register(:inventory) { |sender, message| events << "inventory<-#{sender}:#{message}" }
      mediator.register(:payment) { |sender, message| events << "payment<-#{sender}:#{message}" }

      mediator.send(:payment, :inventory, :paid)
      mediator.send(:inventory, :payment, :reserved)

      expected = ['inventory<-payment:paid', 'payment<-inventory:reserved']
      raise 'Mediator failed' unless events == expected

      true
    end
  end
end

RubyPatterns::Mediator.run if $PROGRAM_NAME == __FILE__
